src/ocean/text/convert/Formatter.d

```diff
@@ -962,3 +984,35 @@ unittest
     static assert(canSwitchOn!(immutable int));
     static assert(!canSwitchOn!(real));
 }
+
+private template isInputRange (R)
+{
+    enum bool isInputRange =
```

Comment:
This is partially correct, but it's not checking the type of the return value of `empty` and `front`.

For example:

https://github.com/dlang/phobos/blob/8a38e6e56218923e04b623362daa4eb78f9eb442/std/range/primitives.d#L172-L177

Since `ReturnType` is complicated and requires copying a lot of code from Phobos, I suggest instead to rewrite `isInputRange` just slightly:

```D
enum bool isInputRange(R) =
    is(typeof(R.init) == R)
    && is(typeof((R r) { return r.empty; }(R.init)) == bool)
    && is(typeof((return ref R r) => r.front))
    && !is(typeof((R r) { return r.front; }(R.init)) == void)
    && is(typeof((R r) => r.popFront));
```

And also the two unittests from Phobos can be copied over:

```D

///
@safe unittest
{
    struct A {}
    struct B
    {
        void popFront();
        @property bool empty();
        @property int front();
    }

    static assert(!isInputRange!A);
    static assert( isInputRange!B);
    static assert( isInputRange!(int[]));
    static assert( isInputRange!(char[]));
    static assert(!isInputRange!(char[4]));
    static assert( isInputRange!(inout(int)[]));

    static struct NotDefaultConstructible
    {
        @disable this();
        void popFront();
        @property bool empty();
        @property int front();
    }
    static assert( isInputRange!NotDefaultConstructible);

    static struct NotDefaultConstructibleOrCopyable
    {
        @disable this();
        @disable this(this);
        void popFront();
        @property bool empty();
        @property int front();
    }
    static assert(isInputRange!NotDefaultConstructibleOrCopyable);

    static struct Frontless
    {
        void popFront();
        @property bool empty();
    }
    static assert(!isInputRange!Frontless);

    static struct VoidFront
    {
        void popFront();
        @property bool empty();
        void front();
    }
    static assert(!isInputRange!VoidFront);
}

@safe unittest
{
    import std.algorithm.comparison : equal;

    static struct R
    {
        static struct Front
        {
            R* impl;
            @property int value() { return impl._front; }
            alias value this;
        }

        int _front;

        @property bool empty() { return _front >= 3; }
        @property auto front() { return Front(&this); }
        void popFront() { _front++; }
    }
    R r;

    static assert(isInputRange!R);
    assert(r.equal([ 0, 1, 2 ]));
}
```

The tests pass with this.

@Geod24 I've just noticed something with my suggestion:

```D
static struct S
{
    @disable this();
    void popFront();
    @property bool empty();
    @property int front();

}

// == 'bool'. But how come 'S.init' doesn't error?
pragma(msg, typeof((S r) { return r.empty; }(S.init)));
```

It seems that some semantic checks are not done within the `typeof()` expression? I wonder if it's on purpose, and if so then `isInputRange` in Phobos could be simplified by getting rid of the `ReturnType` instantiation..

Reply:
I also saw the source code and test code of Phobos, which was strange. It didn't work, so I wrote it using the least.

Reply:
@AndrejMitrovic : I'm not sure what you are referring to here. What error do you expect ?

Reply:
Why not these?
``` D
    static assert( isInputRange!(int[]));
    static assert( isInputRange!(char[]));
    static assert(!isInputRange!(char[4]));
    static assert( isInputRange!(inout(int)[]));
```

Reply:
Only `char[4]` is not an input range. The rest are input ranges.

`char[4]` is not an input range because you cannot call `popFront` on it.